text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+
+The remaining behaviour of the function is equivalent to the `read_freeze` intrinsic.
+
+### Example Alternative Implementation
+
+If an implementation decides to define `MaybeUninit::freeze` (or equivalent) as the compiler intrinsic, it is possible to implement `core::ptr::read_freeze` as follows:
+```rust 
+pub unsafe fn read_freeze(ptr: *const T) -> T{
+    struct AlignedBytes([u8; core::mem::size_of::<T>()], [T;0]);
+    
+    let val = ptr.cast::<MaybeUninit<AlignedBytes>>().read();
+    
+    core::mem::transmute(val.freeze())
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawbacks that have been identified so far:
+* It is potentially [considered desireable](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Arguments.20for.20freeze/near/377333420) to maintain the property that sound (e.g. correct) code cannot meaningfully read uninitialized memory
+    * It is generally noted that safe/unsafe is not a security or privilege boundary, and it's fully possible to write unsound code (either deliberately or inadvertanly) that performs the read. If the use of uninitialized memory is within the threat model of a library that, for example, handles cryptographic secrets, that library should take additional steps to santize memory that contains those secrets.
+    * Undefined behaviour does not prevent malicious code from accessing any memory it physically can
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* The intrinsic could be defined as a by-value operation used by `MaybeUninit::freeze`, instead of the read operation used by `core::ptr::read_freeze`
+    * As noted above, it is believed that the two intrinsic choices are functionally identical, and thus the choice between them is completely arbitrary.
+* Either one of the two functions could be provided on their own
+    * Both functions are provided for maximum flexibility, and can be defined in terms of each other. The author does not believe there is significant drawback to providing both functions instead of just one
+* An in-place, mutable `freeze` could be offered, e.g. `MaybeUninit::freeze(&mut self)`
+    * While this function would seem to be a simple body that llvm could replace with a runtime no-op, in reality it is possible for [virtual memory](https://man7.org/linux/man-pages/man2/madvise.2.html#MADV_FREE) that has been freshly allocated and has not been written to exhibit properties of uninitialized memory (rather than simply being an abstract phenomenon the compiler tracks that disappears at runtime). Thus, such an operation would require a potentially expensive in-place copy. Until such time as an optimized version is available, we should avoid defining the in-place version, and require users to spell it explicitly as `*self = core::mem::replace(&mut self, uninit()).freeze()`.
```

Comment:
I don't think I totally understand the performance considerations here. What is an "in place" copy? Why is it more expensive than a move?

Reply:
It's not more expensive than a move (except as to llvm having more freedom to elide it), but I'm guessing a move looks more like it's "Doing something", whereas a user might assume that `MaybeUninit::freeze(&mut self)` is really a runtime no-op, and complain it's O(n) on their massive structure.

Reply:
Just rewording how I interpreted this, to verify I'm understanding correctly: since virtual memory may be uninitialised but still copy-on-write (for example, from a forked process), there are cases where a copy would still happen even if it's not explicit in the in-place signature, and so it's probably easier to just treat as a read anyway?

Reply:
I mean, I personally think that "user isn't capable of reading documentation" is a compelling case for advocating for a particular API. `slice::reverse` is also an in-place mut method but users don't have issues understanding that it takes linear time. I know there's a better justification, so, I'd personally avoid that kind of argument.

Reply:
Putting on my Project Safe Transmute hat, I'd like to advocate for alternatively or additionally providing an in-place intrinsic. Libraries like `zerocopy` and `bytemuck` almost exclusively work with *references* to values, and those values are potentially DST. The proposed by-value intrinsic does not lend itself to working on DSTs. At the very least, I'd like this RFC to propose a snippet that used the by-value intrinsic to implement a by-ref routine that works on slice DSTs. 

That said, the safety considerations of implementing a general, in-place `freeze` are quite subtle. Zerocopy, for instance, requires that every `unsafe` block is convincingly proven to be sound with citations to authoritative documentation. I think we're very far from having an abstract machine for Rust that is sufficiently well specified that zerocopy could robustly justify the operation described in https://github.com/rust-lang/rfcs/pull/3605#discussion_r1550845642. I would far prefer this fundamental operation to be provided by the standard library, than for it to be re-invented ad hoc by crates.